### PR TITLE
Add json worker to monaco editor

### LINKED ISSUE #3132: Add json worker and language implementation in @osd/monaco
**Is your feature request related to a problem? Please describe.**

We are developing index state management and using `monaco-editor` to edit JSON content, however I find the monaco instance does not support json language.
image

Our product want to enhance the json editor with the capability to validate the properties of json in different case, e.g. Validate the index settings and give a hint on line if the property is not supported.
image

**Describe the solution you'd like**

Add the support of the language of json and add `monaco.languages.json` object to enable plugins to add [DiagnosticsOptions](https://microsoft.github.io/monaco-editor/api/interfaces/monaco.languages.json.DiagnosticsOptions.html).

**Additional context**

You can find the JSON validation playground [here](https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-configure-json-defaults)

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
Signed-off-by: suzhou <suzhou@amazon.com>

### Description
1. Add a new entry in webpack.config.js to build `json.editor.worker.js`.
 
### Issues Resolved
#3132 
 
### Check List
- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
  - [x] `yarn test:ftr`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by SuZhou-Joe** [comment]
#3132

**Comment by SuZhou-Joe** [comment]
#3132

**Comment by AMoo-Miki** [comment]
@SuZhou-Joe Can you please look into the test failures?

**Comment by SuZhou-Joe** [comment]
> @SuZhou-Joe Can you please look into the test failures?

Yes, the monaco implementation has side effect and can not be import twice in a repo, change to only add a json worker.

**Comment by SuZhou-Joe** [comment]
Build and test / Run backwards compatibility tests (osd-2.1.0) (pull_request). Caused by flaky test and passed on this flow https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3424/commits/d80b42e9f0ef4ff2dfc176bd7217394288decc9e .

**Comment by AMoo-Miki** [comment]
Question: Since this is working on top of `monaco-next`, how will it impact plugins that continue to use `monaco`?

Also, can you please rebase on top of `main` and resolve conflicts?

**Comment by SuZhou-Joe** [comment]
> Question: Since this is working on top of `monaco-next`, how will it impact plugins that continue to use `monaco`?
> 
> Also, can you please rebase on top of `main` and resolve conflicts?

Yes, after a quick try to add "@osd/monaco-next", we found some issues happening on other plugins, which indicate the monaco implementation has some side effect. So the PR now only adds a json worker and move the language implementation into IM plugin until the 0.20.0 research and migration are complete.

**Comment by SuZhou-Joe** [comment]
> Question: Since this is working on top of `monaco-next`, how will it impact plugins that continue to use `monaco`?
> 
> Also, can you please rebase on top of `main` and resolve conflicts?

Yes, after a quick try to add "@osd/monaco-next", we found some issues happening on other plugins, which indicate the monaco implementation has some side effect. So the PR now only adds a json worker and move the language implementation into IM plugin until the 0.20.0 research and migration are complete.

**Review by joshuarrrr — COMMENTED** [review]
@SuZhou-Joe Does this PR obsolete https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3133? If so, can you close that one?

**Review by ananzh — APPROVED** [review]
LGTM

NIT: I think you probably use need to update the README. Currently, it still only has intro of XJSON

```
This folder contains the language definitions for XJSON used by the Monaco editor.
```

**Comment by ananzh** [comment]
@SuZhou-Joe 
I see you have a README here: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3133
Are these duplicate? If so, could you add README from 3133 and close 3133? Thank you.

**Comment by SuZhou-Joe** [comment]
> @SuZhou-Joe I see you have a README here: #3133 Are these duplicate? If so, could you add README from 3133 and close 3133? Thank you.

Done for this.

**Comment by SuZhou-Joe** [comment]
> LGTM
> 
> NIT: I think you probably use need to update the README. Currently, it still only has intro of XJSON
> 
> ```
> This folder contains the language definitions for XJSON used by the Monaco editor.
> ```

Done for this.

File: packages/osd-monaco/README.md [hunk 1]
```diff
@@ -2,4 +2,6 @@
 
 A customized version of monaco that is automatically configured the way we want it to be when imported as `@osd/monaco`. Additionally, imports to this package are automatically shared with all plugins using `@osd/ui-shared-deps`.
 
-Includes custom xjson language support. The `opensearch_ui_shared` plugin has an example of how to use it, in the future we will likely expose helpers from this package to make using it everywhere a little more seamless.
\ No newline at end of file
+Includes custom xjson language support. The `opensearch_ui_shared` plugin has an example of how to use it, in the future we will likely expose helpers from this package to make using it everywhere a little more seamless.
+
+Add json language worker export.
```
Comment (joshuarrrr) [inline packages/osd-monaco/README.md:7]:
```suggestion
Includes json language worker support.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -182,6 +182,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - [Multi DataSource] UX enhacement on index pattern management stack ([#2505](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2505))
 - [Multi DataSource] UX enhancement on Data source management stack ([#2521](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2521))
 - [Multi DataSource] UX enhancement on Update stored password modal for Data source management stack ([#2532](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2532))
+- [Monaco editor] add json worker ([#3424](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3424))
```
Comment (joshuarrrr) [inline CHANGELOG.md:185]:
```suggestion
- [Monaco editor] Add json worker support ([#3424](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3424))
```

File: packages/osd-monaco/src/xjson/README.md [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # README
 
-This folder contains the language definitions for XJSON used by the Monaco editor.
+This folder contains the language definitions for XJSON and language worker of JSON used by the Monaco editor.
```
Comment (joshuarrrr) [inline packages/osd-monaco/src/xjson/README.md:3]: revert this - there's no json-specific files or functions in this `xjson` directory.
```suggestion
This folder contains the language definitions for XJSON used by the Monaco editor.
```

**Review by ananzh — APPROVED** [review]

**Comment by SuZhou-Joe** [comment]
> @SuZhou-Joe Does this PR obsolete #3133? If so, can you close that one?

Sure, done for this.

**Comment by ananzh** [comment]
hey @SuZhou-Joe ,

Today is code freeze of 2.6. We have to move this to 2.7.

**Comment by ananzh** [comment]
The pull request adds a new plugin called monaco_editor that includes the necessary code to embed the Monaco editor into OpenSearch-Dashboards. When I reviewed it, I am stuck at understanding 1)whether the changes involve updating various AngularJS controllers, services, and UI templates to support the new editor, as well as adding new CSS styles and configuration files. 2) how to test thoroughly.

**Comment by SuZhou-Joe** [comment]
>

**Comment by SuZhou-Joe** [comment]
> The pull request adds a new plugin called monaco_editor that includes the necessary code to embed the Monaco editor into OpenSearch-Dashboards. When I reviewed it, I am stuck at understanding 1)whether the changes involve updating various services, and UI templates to support the new editor, as well as adding new CSS styles and configuration files. 2) how to test thoroughly.

1. Nope, it only involves a web worker to analyze the content, exactly the same as xjson worker.
2. Monaco itself has done testing on this worker.

**Comment by ashwin-pc** [comment]
@SuZhou-Joe The code change looks fine to me, however im struggling to reproduce the original and modified screens that you are showing. Are there any steps i can use to validate the change that you are making? 

There are two instances of this editor being used in OSD:
1. Timeline visualizations
2. Vega lite spec inspector

In both instances, i do not see the syntax highlighting that this change is supposed to introduce.

**Review by ashwin-pc — APPROVED** [review]
Sorry for the delay. This change looks good to me. 

Note: Just want to call out for future reference that this PR alone does not add JSON support to the monaco editor in OSD, it only adds a json worker that the ISM plugin will use to load the json language support code it has within the plugin. This monaco editor in OSD will need a rework to properly support other languages, and this PR does not change that. Approving to unblock ISM json features.

**Review by joshuarrrr — APPROVED** [review]

**Pull request merged**